### PR TITLE
feat: multi-data connections support for ae

### LINKED ISSUE #1981: Multi-data connections support for AE
## Description
To be filled

## Has the Feature been Requested Before?
Please link any other

### LINKED ISSUE #2256: service init use enckey before it be setting on main
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar issues.

### What happened

In [services/init.go](https://github.com/apache/incubator-devlake/tree/main/services/init.go)
it is init before [main.go](https://github.com/apache/incubator-devlake/tree/main/main.go)
But the encKey are being initialization at the main()
The services init needed encKey for connction.
So it will visit encKey before it initialization .

### What you expected to happen

services init() after encKey initialization.

### How to reproduce

run init() in main() to control init timing.

### Anything else

_No response_

### Version

main

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by mappjzc**
# Summary
Add multi-data support for AE
Mv randomstr to core
Add TestConnection for AE

### Does this close any open issues?
close #1981 
close #2256 

### Screenshots
image
image
image
image
image
image
image

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: plugins/ae/api/connection.go [hunk 1]
```diff
@@ -18,88 +18,165 @@ limitations under the License.
...
+func Init(config *viper.Viper, logger core.Logger, database *gorm.DB) {
+	basicRes := helper.NewDefaultBasicRes(config, logger, database)
+	vld := validator.New()
+	connectionHelper = helper.NewConnectionHelper(
+		basicRes,
+		vld,
+	)
+}
+
+func GetSign(query url.Values, appId, secretKey, nonceStr, timestamp string) string {
```
Comment (klesh) [inline plugins/ae/api/connection.go:53]: I think a better place for this function should be `models/connection.go`

File: plugins/ae/api/connection.go [hunk 2]
```diff
@@ -18,88 +18,165 @@ limitations under the License.
...
+		nonceStr := core.RandLetterBytes(8)
+		timestamp := fmt.Sprintf("%v", time.Now().Unix())
+		sign := GetSign(req.URL.Query(), appId, secretKey, nonceStr, timestamp)
+		req.Header.Set("x-ae-app-id", appId)
+		req.Header.Set("x-ae-timestamp", timestamp)
+		req.Header.Set("x-ae-nonce-str", nonceStr)
+		req.Header.Set("x-ae-sign", sign)
+		return nil
+	})
+	res, err := apiClient.Get("projects/", nil, nil)
```
Comment (klesh) [inline plugins/ae/api/connection.go:110]: Try remove trailing /, should not be expecting 404

File: plugins/ae/models/migrationscripts/archived/connection.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package archived
+
+import "github.com/apache/incubator-devlake/plugins/helper"
+
+type AeConnection struct {
+	helper.RestConnection `mapstructure:",squash"`
```
Comment (klesh) [inline plugins/ae/models/migrationscripts/archived/connection.go:23]: Do not refer struct outside of migrationscripts, they might be modified in future

File: plugins/ae/tasks/api_client.go [hunk 4]
```diff
@@ -18,85 +18,31 @@ limitations under the License.
...
 	apiClient, err := helper.NewApiClient(endpoint, nil, 0, proxy, taskCtx.GetContext())
 	if err != nil {
 		return nil, err
 	}
 	apiClient.SetBeforeFunction(func(req *http.Request) error {
-		nonceStr := RandString(8)
+		nonceStr := core.RandLetterBytes(8)
 		timestamp := fmt.Sprintf("%v", time.Now().Unix())
-		sign := getSign(req.URL.Query(), appId, secretKey, nonceStr, timestamp)
+		sign := api.GetSign(req.URL.Query(), appId, secretKey, nonceStr, timestamp)
```
Comment (klesh) [inline plugins/ae/tasks/api_client.go:45]: It makes no sense, for a `tasks` package to use sth from `api` package.
`api` package should be always sitting on top of everything except `test cases`

File: plugins/ae/tasks/task_data.go [hunk 5]
```diff
@@ -20,8 +20,9 @@ package tasks
 import "github.com/apache/incubator-devlake/plugins/helper"
 
 type AeOptions struct {
-	ProjectId int
-	Tasks     []string `json:"tasks,omitempty"`
+	ConnectionId uint64 `json:"connectionId"`
+	ProjectId    int
+	Tasks        []string `json:"tasks,omitempty"`
```
Comment (klesh) [inline plugins/ae/tasks/task_data.go:25]: not needed anymore, please remove this field

File: plugins/ae/ae.go [hunk 6]
```diff
@@ -82,20 +101,25 @@ func (plugin AE) RootPkgPath() string {
...
+	return []migration.Script{
+		new(migrationscripts.InitSchemas),
+		new(migrationscripts.UpdateSchemas20220615),
+	}
 }
 
 func (plugin AE) ApiResources() map[string]map[string]core.ApiResourceHandler {
 	return map[string]map[string]core.ApiResourceHandler{
-		"test": {
+		"test/:connectionId": {
```
Comment (klesh) [inline plugins/ae/ae.go:112]: No, connections/test is used by config-ui to test if the connection works or not BEFORE saving it. 
At that point, no :connectionId yet

File: plugins/ae/api/connection.go [hunk 7]
```diff
@@ -18,88 +18,133 @@ limitations under the License.
...
+POST /plugins/ae/connections
 */
-func ListConnections(input *core.ApiResourceInput) (*core.ApiResourceOutput, error) {
-	// RETURN ONLY 1 SOURCE (FROM ENV) until multi-connection is developed.
-	v := config.GetConfig()
+func PostConnections(input *core.ApiResourceInput) (*core.ApiResourceOutput, error) {
 	connection := &models.AeConnection{}
-
-	err := helper.EncodeStruct(v, connection, "env")
+	fmt.Printf("%+v\r\n", connectionHelper)
```
Comment (klesh) [inline plugins/ae/api/connection.go:98]: Either use logger or remove it .

File: plugins/ae/api/connection.go [hunk 8]
```diff
@@ -18,88 +18,141 @@ limitations under the License.
...
+	vld = validator.New()
+	connectionHelper = helper.NewConnectionHelper(
+		basicRes,
+		vld,
+	)
 }
 
 /*
-PATCH /plugins/ae/connections/:connectionId
+GET /plugins/ae/test/:connectionId
```
Comment (klesh) [inline plugins/ae/api/connection.go:51]: The comment doesn't seem correct.

**Review by klesh — COMMENTED** [review]
Good job, one last small problem to solve then we can merge

**Review by likyh — APPROVED** [review]
LGTM

**Pull request merged**